feat: add multiple features and fix

coanor opened the pull request:
- Enable list metrics in vendor packages(`--with-vendor`)
    - Why? Some vendor packages also export metrics, we need to export them for full metric documents.
- List labels of metrics, if no labels, output `N/A`(#37)
- Fix panic on `--add-help` if metric got no help info, and also if no help output `N/A`
- Out JSON with pretty indents
- Output markdown table(#38)
- Avoid duplicated metric name(#40)

Close #37/#38/#40.

coanor committed:
feat: add multiple features

- Enable list metrics in vendor packages(--with-vendor)
- List labels of metrics(if no labels, output `N/A')
- Fix panic on --add-help if metric got no help info

coanor committed:
fix: fix typo

coanor commented:
The output example seems like this:

```shell
TYPE                NAME                    LABELS       HELP
COUNTER             http_api_signed_total   api,method   API signature count
COUNTER             http_api_dropped_total  api,method   API request dropped when sinker rule match failed
```

Labels are `,` splited string. If there're const labels, the label displayed as:

- `some_const_label=some_value`: The const label got a string literal value
- `some_const_label=?`: While the const label value are not string literal(such as from `fmt.Sprintf()` or `someModule.SomeValue`)

And the JSON format like this:

```json
[
  {
    "Name": "http_api_dropped_total",
    "Help": "API request dropped when sinker rule match failed",
    "Type": "COUNTER",
    "Filename": "/path/to/metrics.go",
    "Labels": [
      "api",
      "method"
    ],
    "Line": 105,
    "Column": 3
  },
  {
    "Name": "http_api_signed_total",
    "Help": "API signature count",
    "Type": "COUNTER",
    "Filename": "/path/to/metrics.go",
    "Labels": [
      "api",
      "method"
    ],
    "Line": 118,
    "Column": 3
  }
]
```

The markdown table seems like this:

```markdown
| TYPE    | NAME                     | LABELS       | HELP                                              |
| ---     | ---                      | ---          | ---                                               |
| COUNTER | `http_api_signed_total`  | `api,method` | API signature count                               |
| COUNTER | `http_api_dropped_total` | `api,method` | API request dropped when sinker rule match failed |
```

yeya24 commented:
Thanks for the good work! @coanor Are you able to add some unit tests about the labels feature?

coanor committed:
feat: add markdown table output

- Fix duplicated metric name output

coanor commented:
> Thanks for the good work! @coanor Are you able to add some unit tests about the labels feature?

I find the PR can't find labels within `NewDesc`, I need some time to refactor current `parseXXX` functions. And current PR closed, I'll reopen it after `variableLabels` and `constLabels` able to parse.

coanor committed:
add label and const label discovery

coanor committed:
add test on list labels

coanor committed:
update test cases

coanor commented:
> Thanks for the good work! @coanor Are you able to add some unit tests about the labels feature?

I've add label discovery within `NewDesc` and basic test case, but it seems that the original case within *promlinter_test.go* failed, there missing a issue in current MR, currently I have no idea about what happend.

Will you give me some hint about why the case failed?

coanor committed:
feat: add print option --add-module

Print position(file name or path) in italic under markdown table output

yeya24 commented:
It seems like the error case below was missing.

```
	if issues[2].Metric != "test_metric_total" && issues[2].Text != `no help text` {
		t.Fatal()
	}
```

coanor committed:
save

coanor committed:
fix unit test fail

coanor commented:
> It seems like the error case case below was missing.
> 
> ```
> 	if issues[2].Metric != "test_metric_total" && issues[2].Text != `no help text` {
> 		t.Fatal()
> 	}
> ```

I've refactor and fixed the testings within *promlinter_test.go*, It seems that the issue's order not the key concerning.

coanor committed:
remove debugging logging

yeya24 commented on promlinter.go:
```diff
@@ -254,19 +291,44 @@ func (v *visitor) parseCallerExpr(call *ast.CallExpr) ast.Visitor {
+	}
+
+	if !opts.helpSet {
+		currentMetric.Help = nil
```
We don't need this condition. We cannot set String to `nil` I guess.
Line 327 should be enough to check if help is set

coanor replied:
`currentMetric.Help` is type `*string`, we just assign nil to mean "There is no help available". If we set a point to empty string(`""`, here is line 327's `opts.help`, it's default value is `""`), some exists test cases will fail(*empty string still a valid help, but not make sense.*).

yeya24 commented on promlinter.go:
```diff
@@ -395,17 +483,50 @@ func (v *visitor) parseOptsExpr(n ast.Node) (*opt, *string) {
+		if ok {
+			metricOption.labels = append(metricOption.labels, strings.Trim(label.Value, `"`))
+			continue
+			// go down
```
Do we need this comment?

coanor replied:
No, this may a noise comment, just removed.

yeya24 commented:
> I've refactor and fixed the testings within promlinter_test.go, It seems that the issue's order not the key concerning.

I am ok to refactor the tests but I think it is still good to return ordered (at least order can be determined) issues from `RunLint()`.

coanor committed:
remove noise comment.

coanor committed:
save

coanor committed:
remove sort on result issue.

coanor commented:
> > I've refactor and fixed the testings within promlinter_test.go, It seems that the issue's order not the key concerning.
> 
> I am ok to refactor the tests but I think it is still good to return ordered (at least order can be determined) issues from `RunLint()`.

About the sort, I think we do not need to sort the result issue, because they are sorted already according to the AST traverse(And I removed the trailing `sort` on issues(fd4631db6cd98966c40b2d6b2d8740644d30d55f)).

For example, the raw issue list is(from *testdata/testdata.go*)

```
test_metric_name: ./testdata/testdata.go:23:3
test_metric_total: ./testdata/testdata.go:32:3
foo: ./testdata/testdata.go:39:30
foo_bar_total: ./testdata/testdata.go:78:8
kube_test_metric_count: ./testdata/testdata.go:92:8
test_histogram_duration_seconds: ./testdata/testdata.go:95:27
kube_daemonset_labels: ./testdata/testdata.go:119:4
```

After sort(based on `pos.String()`), the ording changed(and is not the AST order):

```
kube_daemonset_labels: ./testdata/testdata.go:119:4
test_metric_name: ./testdata/testdata.go:23:3
test_metric_total: ./testdata/testdata.go:32:3
foo: ./testdata/testdata.go:39:30
foo_bar_total: ./testdata/testdata.go:78:8
kube_test_metric_count: ./testdata/testdata.go:92:8
test_histogram_duration_seconds: ./testdata/testdata.go:95:27
```

Because string `a.go:123 < a.go:99`, and that *not* the order we need：

```go
import (
  T "testing"

  "github.com/stretchr/testify/assert"
)

func TestStringOrder(t *T.T) {
	assert.True(t, "a.go:123" < "a.go:99")  // Pass
}
```

yeya24 committed:
Merge branch 'master' into master

yeya24 commented:
Sorry for taking it for so long. I am going to merge it. Great work! Thanks
